### PR TITLE
Refactor `CostDistribution` and update docs

### LINKED ISSUE #133: cost_utils.CostDistribution Incorrect Attribute docstring
the attributes outlined in the attribute docstring do not exist and are misleading as to how the class init works

### PR DESCRIPTION
**Pull request by BenTaylor-TfN**
Describe Changes
---
Fixes #133 

* Refactored `CostDistribution` making some implicitly private attributes private.
* Added documentation for the constructor
* Updated tests to reflect refactor

Task Checklist
----
- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Have integration tests been added (if applicable, to test modules of functionality)
- [ ] Updated RELEASE.md to reflect changes in PR
- [ ] Have new dependencies been added?
  - [ ] Have they been added to either `requirements.txt` or `requirements_dev.txt`.
  - [ ] Have they been added to `pyproject.toml`

### CODE REVIEW

File: src/caf/toolkit/cost_utils.py [hunk 1]
```diff
@@ -6,98 +6,116 @@
...
     """Distribution of cost values between variable bounds.
 
-    Attributes
+    Alternate constructors are available in the See Also section
+
+    Parameters
     ----------
     df:
-        The raw Pandas DataFrame containing the cost distribution data.
+        A DataFrame containing the cost distribution data. Must have columns
```
Comment (Kieran-Fishwick-TfN) [inline src/caf/toolkit/cost_utils.py:32]:
```suggestion
        A DataFrame containing the binned cost distribution data. Must have columns
```

File: src/caf/toolkit/cost_utils.py [hunk 2]
```diff
@@ -6,98 +6,116 @@
...
+        named: `min_col`, `max_col`, `avg_col`, `trips_col`.
 
-    min_vals:
-        Minimum values of the cost distribution bin edges.
+    min_col:
+        The name of the columns in `df` that contains the minimum bound
```
Comment (Kieran-Fishwick-TfN) [inline src/caf/toolkit/cost_utils.py:36]:
```suggestion
        The name of the columns in `df` that contains the lower bin edge
```

File: src/caf/toolkit/cost_utils.py [hunk 3]
```diff
@@ -6,98 +6,116 @@
...
+        value for each row.
 
-    max_vals:
-        Maximum values of the cost distribution bin edges.
+    max_col:
+        The name of the columns in `df` that contains the maximum bound
```
Comment (Kieran-Fishwick-TfN) [inline src/caf/toolkit/cost_utils.py:40]:
```suggestion
        The name of the columns in `df` that contains the upper bin edge
```

File: src/caf/toolkit/cost_utils.py [hunk 4]
```diff
@@ -6,98 +6,116 @@
...
+        value for each row.
 
-    bin_edges:
-        Bin edges for the cost distribution.
-
-    avg_vals:
-        Average values for each of the cost distribution bins.
+    avg_col:
+        The name of the columns in `df` that contains the middle bound
```
Comment (Kieran-Fishwick-TfN) [inline src/caf/toolkit/cost_utils.py:44]:
```suggestion
        The name of the columns in `df` that contains the centre of the bin
```

File: src/caf/toolkit/cost_utils.py [hunk 5]
```diff
@@ -6,98 +6,116 @@
...
-
-    # Default arguments
-    min_col: str = "min"
-    max_col: str = "max"
-    avg_col: str = "ave"
-    trips_col: str = "trips"
-    weighted_avg_col: Optional[str] = "weighted_ave"
-
     # Ideas
     units: str = "km"
```
Comment (MattBuckley-TfN) [inline src/caf/toolkit/cost_utils.py:64]: Does this want to be a parameter that instead of being hardcoded?

File: src/caf/toolkit/cost_utils.py [hunk 6]
```diff
@@ -6,98 +6,116 @@
...
+    def __init__(
+        self,
+        df: pd.DataFrame,
+        min_col: str = "min",
+        max_col: str = "max",
+        avg_col: str = "avg",
+        trips_col: str = "trips",
+        weighted_avg_col: Optional[str] = None,
+    ):
+        self.df = df
```
Comment (MattBuckley-TfN) [inline src/caf/toolkit/cost_utils.py:75]: Should the dataframe be copied (shallowly) to avoid people potentially manipulating it after passing it?

File: src/caf/toolkit/cost_utils.py [hunk 7]
```diff
@@ -6,98 +6,116 @@
...
+
+        # Keep as private. These shouldn't be needed outside of this class
+        self.__min_col = min_col
```
Comment (MattBuckley-TfN) [inline src/caf/toolkit/cost_utils.py:78]: Why 2 `_` underscores I thought 1 was how private attribute were defined?

File: src/caf/toolkit/cost_utils.py [hunk 8]
```diff
@@ -6,98 +6,116 @@
...
+        self.__max_col = max_col
+        self.__avg_col = avg_col
+        self.__trips_col = trips_col
+
+        if weighted_avg_col is None:
+            weighted_avg_col = "weighted_avg"
+        self.__weighted_avg_col = weighted_avg_col
```
Comment (MattBuckley-TfN) [inline src/caf/toolkit/cost_utils.py:85]: This could just be the parameters default value since it doesn't do anything special if given None

File: src/caf/toolkit/cost_utils.py [hunk 9]
```diff
@@ -262,14 +285,21 @@ def from_data(
         # Covert data into instance of this class
         df = pd.DataFrame(
             {
-                cls.min_col: bin_edges[:-1],
-                cls.max_col: bin_edges[1:],
-                cls.avg_col: (np.array(bin_edges[:-1]) + np.array(bin_edges[1:])) / 2,
-                cls.trips_col: distribution,
-                cls.weighted_avg_col: averages,
+                "min": bin_edges[:-1],
```
Comment (MattBuckley-TfN) [inline src/caf/toolkit/cost_utils.py:288]: Is it best practise not to access private class attributes from classmethods? I thought you were fine to access any private class functionality from any methods within the class?

File: src/caf/toolkit/cost_utils.py [hunk 10]
```diff
@@ -353,15 +383,17 @@ def from_file(
         cost_distribution:
             An instance containing the data at filepath.
         """
+        # Validate the path
         if not os.path.isfile(filepath):
             raise ValueError(f"'{filepath}' is not the location of a file.")
```
Comment (MattBuckley-TfN) [inline src/caf/toolkit/cost_utils.py:388]: Would a `FileNotFoundError` be more appropriate here?

File: src/caf/toolkit/cost_utils.py (hunk 6 above)
Comment (MattBuckley-TfN) [reply src/caf/toolkit/cost_utils.py:75]: Looks like df is copied during the reindex process within `_validate_df_col_names` anyway, so probably unnecessary here.

**Review by MattBuckley-TfN — APPROVED** [review]
All functionality looks good just got some clarification questions around why you've done things a particular way.

File: src/caf/toolkit/cost_utils.py (hunk 5 above)
Comment (BenTaylor-TfN) [reply src/caf/toolkit/cost_utils.py:64]: Weird hangover for development - will comment out
```suggestion
    # units: str = "km"
```

File: src/caf/toolkit/cost_utils.py (hunk 7 above)
Comment (BenTaylor-TfN) [reply src/caf/toolkit/cost_utils.py:78]: `__` uses name mangling, so it's even harder to reference the private items e.g. `instance.__min_col` won't work, instead `instance._CostDistribution__min_col` will have to be used.

In this case I **never** want someone to refer to specific colums in the df and instead use the attributes provided to get the data instead.

Comment (MattBuckley-TfN) [reply src/caf/toolkit/cost_utils.py:78]: Oh right, I'll keep that in mind for future thanks
